File: SOLID/code-examples/chapter-14.cpp
// FILE: chapter-14-clean-code.cpp
// Chapter 14 — Clean Code Basics — Beginner/Intermediate

#include <iostream>
#include <string>
#include <vector>
#include <memory>
#include <stdexcept>
#include <regex>
#include <iomanip>

// BAD EXAMPLE - Messy, unreadable code
class u {
public:
    std::string n;
    std::string e;
    int a;
    bool v;
    
    u(std::string n, std::string e, int a) : n(n), e(e), a(a), v(false) {}
    
    bool chk() {
        if(n.length() < 2) return false;
        if(e.find("@") == std::string::npos) return false;
        if(a < 0 || a > 150) return false;
        return true;
    }
    
    void p() {
        if(chk()) {
            if(a >= 18) {
                std::cout << "Adult: " << n << std::endl;
            } else {
                std::cout << "Minor: " << n << std::endl;
            }
        }
    }
};

// GOOD EXAMPLE - Clean, readable code

// Constants for clarity
namespace UserValidationConstants {
    const int MINIMUM_NAME_LENGTH = 2;
    const int MINIMUM_AGE = 0;
    const int MAXIMUM_AGE = 150;
    const int ADULT_AGE_THRESHOLD = 18;
    const std::string EMAIL_PATTERN = R"(^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$)";
}

// Enum for user types instead of magic numbers
enum class UserType {
    Minor,
    Adult,
    Senior
};

// Custom exception for validation errors
class ValidationError : public std::runtime_error {
public:
    explicit ValidationError(const std::string& message) : std::runtime_error(message) {}
};

// Clean User class with proper encapsulation
class User {
private:
    std::string name;
    std::string email;
    int age;
    bool isVerified;

public:
    /**
     * @brief Construct a new User object with validation
     * 
     * @param name User's full name (must be at least 2 characters)
     * @param email User's email address (must be valid format)
     * @param age User's age (must be between 0 and 150)
     * @throws ValidationError If any input data is invalid
     */
    User(const std::string& name, const std::string& email, int age)
        : name(validateName(name)), email(validateEmail(email)), age(validateAge(age)), isVerified(false) {}

    // Getters with clear names
    const std::string& getName() const { return name; }
    const std::string& getEmail() const { return email; }
    int getAge() const { return age; }
    bool getIsVerified() const { return isVerified; }
    
    // Setter with validation
    void setIsVerified(bool verified) { isVerified = verified; }

    /**
     * @brief Determine the user's category based on age
     * @return UserType category
     */
    UserType getUserType() const {
        if (age >= 65) {
            return UserType::Senior;
        } else if (age >= UserValidationConstants::ADULT_AGE_THRESHOLD) {
            return UserType::Adult;
        } else {
            return UserType::Minor;
        }
    }

    /**
     * @brief Check if all user data is valid according to business rules
     * @return true if all data is valid, false otherwise
     */
    bool isValid() const {
        try {
            validateName(name);
            validateEmail(email);
            validateAge(age);
            return true;
        } catch (const ValidationError&) {
            return false;
        }
    }

    /**
     * @brief Get a display-friendly description of the user
     * @return Formatted string describing the user
     */
    std::string getUserDescription() const {
        if (!isValid()) {
            return "Invalid User";
        }

        std::string ageCategory = getUserTypeString();
        std::string verificationStatus = isVerified ? "Verified" : "Unverified";
        
        return ageCategory + " User: " + name + " (" + verificationStatus + ")";
    }

    /**
     * @brief Display user information to console with proper formatting
     */
    void displayUserInfo() const {
        if (!isValid()) {
            std::cout << "❌ Cannot display invalid user information" << std::endl;
            return;
        }

        std::cout << "👤 User Information:" << std::endl;
        std::cout << "   Name: " << name << std::endl;
        std::cout << "   Email: " << email << std::endl;
        std::cout << "   Age: " << age << std::endl;
        std::cout << "   Type: " << getUserTypeString() << std::endl;
        std::cout << "   Status: " << (isVerified ? "✅ Verified" : "⚠️ Unverified") << std::endl;
    }

private:
    std::string validateName(const std::string& nameValue) const {
        if (nameValue.empty() || nameValue.find_first_not_of(' ') == std::string::npos) {
            throw ValidationError("Name cannot be null or empty");
        }

        std::string trimmedName = trim(nameValue);
        if (trimmedName.length() < UserValidationConstants::MINIMUM_NAME_LENGTH) {
            throw ValidationError("Name must be at least " + 
                std::to_string(UserValidationConstants::MINIMUM_NAME_LENGTH) + " characters long");
        }

        return trimmedName;
    }

    std::string validateEmail(const std::string& emailValue) const {
        if (emailValue.empty()) {
            throw ValidationError("Email cannot be null or empty");
        }

        std::string trimmedEmail = trim(emailValue);
        std::regex emailRegex(UserValidationConstants::EMAIL_PATTERN);
        
        if (!std::regex_match(trimmedEmail, emailRegex)) {
            throw ValidationError("Email must be in valid format");
        }

        // Convert to lowercase for consistency
        std::string lowerEmail = trimmedEmail;
        std::transform(lowerEmail.begin(), lowerEmail.end(), lowerEmail.begin(), ::tolower);
        
        return lowerEmail;
    }

    int validateAge(int ageValue) const {
        if (ageValue < UserValidationConstants::MINIMUM_AGE || ageValue > UserValidationConstants::MAXIMUM_AGE) {
            throw ValidationError("Age must be between " + 
                std::to_string(UserValidationConstants::MINIMUM_AGE) + " and " + 
                std::to_string(UserValidationConstants::MAXIMUM_AGE));
        }

        return ageValue;
    }

    std::string getUserTypeString() const {
        switch (getUserType()) {
            case UserType::Minor: return "Minor";
            case UserType::Adult: return "Adult";
            case UserType::Senior: return "Senior";
            default: return "Unknown";
        }
    }

    std::string trim(const std::string& str) const {
        size_t first = str.find_first_not_of(' ');
        if (first == std::string::npos) return "";
        size_t last = str.find_last_not_of(' ');
        return str.substr(first, (last - first + 1));
    }
};

// Clean Order Processing Example
struct OrderItem {
    std::string productName;
    double price;
    int quantity;
    std::string category;

    double getTotalPrice() const {
        return price * quantity;
    }
};

namespace OrderConstants {
    const double ELECTRONICS_TAX_RATE = 0.1;
    const double BOOK_BULK_DISCOUNT_RATE = 0.05;
    const int BOOK_BULK_QUANTITY_THRESHOLD = 5;
    const double FREE_SHIPPING_THRESHOLD = 100.0;
    const double STANDARD_SHIPPING_COST = 10.0;
    
    const std::string ELECTRONICS_CATEGORY = "ELECTRONICS";
    const std::string BOOKS_CATEGORY = "BOOKS";
}

class OrderCalculator {
public:
    /**
     * @brief Calculate the total order amount including taxes, discounts, and shipping
     * @param orderItems List of items in the order
     * @return Total order amount rounded to 2 decimal places
     * @throws std::invalid_argument If order items are invalid
     */
    double calculateOrderTotal(const std::vector<OrderItem>& orderItems) {
        validateOrderItems(orderItems);

        double subtotal = calculateSubtotal(orderItems);
        double shippingCost = calculateShippingCost(subtotal);
        double totalAmount = subtotal + shippingCost;

        return std::round(totalAmount * 100.0) / 100.0;  // Round to 2 decimal places
    }

private:
    void validateOrderItems(const std::vector<OrderItem>& orderItems) {
        if (orderItems.empty()) {
            throw std::invalid_argument("Order must contain at least one item");
        }

        for (const auto& item : orderItems) {
            if (item.quantity <= 0) {
                throw std::invalid_argument("Invalid quantity for item " + item.productName);
            }

            if (item.price < 0) {
                throw std::invalid_argument("Invalid price for item " + item.productName);
            }

            if (item.productName.empty()) {
                throw std::invalid_argument("Product name cannot be empty");
            }
        }
    }

    double calculateSubtotal(const std::vector<OrderItem>& orderItems) {
        double subtotal = 0.0;

        for (const auto& item : orderItems) {
            double itemTotal = calculateItemTotal(item);
            subtotal += itemTotal;
        }

        return subtotal;
    }

    double calculateItemTotal(const OrderItem& item) {
        double baseTotal = item.getTotalPrice();

        // Apply category-specific pricing rules
        std::string upperCategory = toUpperCase(item.category);
        
        if (upperCategory == OrderConstants::ELECTRONICS_CATEGORY) {
            return applyElectronicsTax(baseTotal);
        } else if (upperCategory == OrderConstants::BOOKS_CATEGORY) {
            return applyBookDiscount(baseTotal, item.quantity);
        } else {
            return baseTotal;
        }
    }

    double applyElectronicsTax(double amount) {
        return amount * (1.0 + OrderConstants::ELECTRONICS_TAX_RATE);
    }

    double applyBookDiscount(double amount, int quantity) {
        if (quantity >= OrderConstants::BOOK_BULK_QUANTITY_THRESHOLD) {
            return amount * (1.0 - OrderConstants::BOOK_BULK_DISCOUNT_RATE);
        }
        
        return amount;
    }

    double calculateShippingCost(double subtotal) {
        return subtotal >= OrderConstants::FREE_SHIPPING_THRESHOLD 
            ? 0.0 
            : OrderConstants::STANDARD_SHIPPING_COST;
    }

    std::string toUpperCase(const std::string& str) {
        std::string result = str;
        std::transform(result.begin(), result.end(), result.begin(), ::toupper);
        return result;
    }
};

// Clean User Service with proper separation of concerns
class UserRepository {
public:
    virtual ~UserRepository() = default;
    virtual void saveUser(const User& user) = 0;
    virtual std::unique_ptr<User> getUserByEmail(const std::string& email) = 0;
};

class EmailService {
public:
    virtual ~EmailService() = default;
    virtual void sendWelcomeEmail(const std::string& email, const std::string& name) = 0;
};

class UserService {
private:
    std::unique_ptr<UserRepository> userRepository;
    std::unique_ptr<EmailService> emailService;

public:
    /**
     * @brief Construct a new User Service with dependencies
     * @param userRepository Repository for user data persistence
     * @param emailService Service for sending emails
     */
    UserService(std::unique_ptr<UserRepository> userRepository, 
                std::unique_ptr<EmailService> emailService)
        : userRepository(std::move(userRepository)), emailService(std::move(emailService)) {
        
        if (!this->userRepository) {
            throw std::invalid_argument("UserRepository cannot be null");
        }
        if (!this->emailService) {
            throw std::invalid_argument("EmailService cannot be null");
        }
    }

    /**
     * @brief Register a new user in the system
     * @param name User's full name
     * @param email User's email address
     * @param age User's age
     * @return Newly created user
     * @throws ValidationError If user data is invalid
     * @throws std::runtime_error If user already exists or save fails
     */
    User registerUser(const std::string& name, const std::string& email, int age) {
        validateUserDoesNotExist(email);

        User newUser = createUser(name, email, age);
        saveUser(newUser);
        sendWelcomeEmail(newUser);

        return newUser;
    }

private:
    void validateUserDoesNotExist(const std::string& email) {
        auto existingUser = userRepository->getUserByEmail(email);
        if (existingUser != nullptr) {
            throw std::runtime_error("User with email " + email + " already exists");
        }
    }

    User createUser(const std::string& name, const std::string& email, int age) {
        try {
            return User(name, email, age);
        } catch (const ValidationError& ex) {
            throw ValidationError("Invalid user data: " + std::string(ex.what()));
        }
    }

    void saveUser(const User& user) {
        try {
            userRepository->saveUser(user);
        } catch (const std::exception& ex) {
            throw std::runtime_error("Failed to save user to database: " + std::string(ex.what()));
        }
    }

    void sendWelcomeEmail(const User& user) {
        try {
            emailService->sendWelcomeEmail(user.getEmail(), user.getName());
        } catch (const std::exception& ex) {
            // Log error but don't fail registration
            std::cout << "Warning: Failed to send welcome email to " << user.getEmail() 
                      << ": " << ex.what() << std::endl;
        }
    }
};

// Demo functions showing clean code in action
void demonstrateCleanUserClass() {
    std::cout << "--- User Management Demo ---" << std::endl;

    try {
        // Create valid users
        User adultUser("Alice Johnson", "alice@example.com", 25);
        User minorUser("Bob Smith", "bob@example.com", 16);

        adultUser.setIsVerified(true);

        adultUser.displayUserInfo();
        std::cout << std::endl;
        minorUser.displayUserInfo();
        std::cout << std::endl;
    } catch (const ValidationError& ex) {
        std::cout << "❌ Error creating user: " << ex.what() << std::endl;
    }

    // Demonstrate validation
    try {
        User invalidUser("X", "invalid-email", -5);
    } catch (const ValidationError& ex) {
        std::cout << "✅ Validation working: " << ex.what() << std::endl;
    }
}

void demonstrateCleanOrderCalculation() {
    std::cout << "\n--- Order Calculation Demo ---" << std::endl;

    std::vector<OrderItem> orderItems = {
        {"Laptop", 999.99, 1, "ELECTRONICS"},
        {"Programming Books", 29.99, 6, "BOOKS"},
        {"Mouse Pad", 9.99, 2, "ACCESSORIES"}
    };

    OrderCalculator calculator;
    double total = calculator.calculateOrderTotal(orderItems);

    std::cout << "Order Summary:" << std::endl;
    for (const auto& item : orderItems) {
        std::cout << "  " << item.productName << ": $" << std::fixed << std::setprecision(2) 
                  << item.price << " x " << item.quantity << " = $" << item.getTotalPrice() << std::endl;
    }
    std::cout << "\nTotal Amount: $" << std::fixed << std::setprecision(2) << total << std::endl;
}

int main() {
    std::cout << "=== Clean Code Demo ===" << std::endl << std::endl;

    demonstrateCleanUserClass();
    demonstrateCleanOrderCalculation();

    std::cout << "\n=== Clean Code Benefits ===" << std::endl;
    std::cout << "✓ Code is easy to read and understand" << std::endl;
    std::cout << "✓ Functions have single responsibilities" << std::endl;
    std::cout << "✓ Names clearly express intent" << std::endl;
    std::cout << "✓ Error handling is explicit and helpful" << std::endl;
    std::cout << "✓ Constants replace magic numbers" << std::endl;
    std::cout << "✓ Code is well-structured and organized" << std::endl;
    std::cout << "✓ Documentation explains purpose and usage" << std::endl;
    std::cout << "✓ RAII and proper resource management" << std::endl;

    return 0;
}

File: SOLID/code-examples/chapter-14.cs
// FILE: chapter-14-clean-code.cs
// Chapter 14 — Clean Code Basics — Beginner/Intermediate

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace CleanCodeExample
{
    // BAD EXAMPLE - Messy, unreadable code
    public class u {
        public string n;
        public string e;
        public int a;
        public bool v;
        
        public bool chk() {
            if(n == null || n.Length < 2) return false;
            if(e == null || !e.Contains("@")) return false;
            if(a < 0 || a > 150) return false;
            return true;
        }
        
        public void p() {
            if(chk()) {
                if(a >= 18) {
                    Console.WriteLine("Adult: " + n);
                } else {
                    Console.WriteLine("Minor: " + n);
                }
            }
        }
    }

    // GOOD EXAMPLE - Clean, readable code

    // Constants for clarity
    public static class UserValidationConstants
    {
        public const int MINIMUM_NAME_LENGTH = 2;
        public const int MINIMUM_AGE = 0;
        public const int MAXIMUM_AGE = 150;
        public const int ADULT_AGE_THRESHOLD = 18;
    }

    // Enum for user types instead of magic numbers
    public enum UserType
    {
        Minor = 0,
        Adult = 1,
        Senior = 2
    }

    // Clean User class with proper encapsulation
    public class User
    {
        private string name;
        private string email;
        private int age;
        private bool isVerified;

        public string Name 
        { 
            get => name; 
            set => name = ValidateName(value); 
        }
        
        public string Email 
        { 
            get => email; 
            set => email = ValidateEmail(value); 
        }
        
        public int Age 
        { 
            get => age; 
            set => age = ValidateAge(value); 
        }
        
        public bool IsVerified 
        { 
            get => isVerified; 
            set => isVerified = value; 
        }

        public UserType UserType => DetermineUserType();

        public User(string name, string email, int age)
        {
            Name = name;
            Email = email;
            Age = age;
            IsVerified = false;
        }

        /// <summary>
        /// Validates all user data according to business rules
        /// </summary>
        /// <returns>True if all data is valid, false otherwise</returns>
        public bool IsValid()
        {
            try
            {
                ValidateName(name);
                ValidateEmail(email);
                ValidateAge(age);
                return true;
            }
            catch (ArgumentException)
            {
                return false;
            }
        }

        /// <summary>
        /// Gets a display-friendly description of the user
        /// </summary>
        /// <returns>Formatted string describing the user</returns>
        public string GetUserDescription()
        {
            if (!IsValid())
            {
                return "Invalid User";
            }

            string ageCategory = UserType == UserType.Adult ? "Adult" : "Minor";
            string verificationStatus = IsVerified ? "Verified" : "Unverified";
            
            return $"{ageCategory} User: {Name} ({verificationStatus})";
        }

        /// <summary>
        /// Displays user information to console with proper formatting
        /// </summary>
        public void DisplayUserInfo()
        {
            if (!IsValid())
            {
                Console.WriteLine("❌ Cannot display invalid user information");
                return;
            }

            Console.WriteLine("👤 User Information:");
            Console.WriteLine($"   Name: {Name}");
            Console.WriteLine($"   Email: {Email}");
            Console.WriteLine($"   Age: {Age}");
            Console.WriteLine($"   Type: {UserType}");
            Console.WriteLine($"   Status: {(IsVerified ? "✅ Verified" : "⚠️ Unverified")}");
        }

        private string ValidateName(string nameValue)
        {
            if (string.IsNullOrWhiteSpace(nameValue))
            {
                throw new ArgumentException("Name cannot be null or empty");
            }

            if (nameValue.Length < UserValidationConstants.MINIMUM_NAME_LENGTH)
            {
                throw new ArgumentException($"Name must be at least {UserValidationConstants.MINIMUM_NAME_LENGTH} characters long");
            }

            return nameValue.Trim();
        }

        private string ValidateEmail(string emailValue)
        {
            if (string.IsNullOrWhiteSpace(emailValue))
            {
                throw new ArgumentException("Email cannot be null or empty");
            }

            if (!emailValue.Contains("@") || !emailValue.Contains("."))
            {
                throw new ArgumentException("Email must be in valid format");
            }

            return emailValue.Trim().ToLower();
        }

        private int ValidateAge(int ageValue)
        {
            if (ageValue < UserValidationConstants.MINIMUM_AGE || ageValue > UserValidationConstants.MAXIMUM_AGE)
            {
                throw new ArgumentException($"Age must be between {UserValidationConstants.MINIMUM_AGE} and {UserValidationConstants.MAXIMUM_AGE}");
            }

            return ageValue;
        }

        private UserType DetermineUserType()
        {
            if (age >= 65)
                return UserType.Senior;
            else if (age >= UserValidationConstants.ADULT_AGE_THRESHOLD)
                return UserType.Adult;
            else
                return UserType.Minor;
        }
    }

    // Clean Order Processing Example
    public class OrderItem
    {
        public string ProductName { get; set; }
        public decimal Price { get; set; }
        public int Quantity { get; set; }
        public string Category { get; set; }

        public decimal TotalPrice => Price * Quantity;
    }

    public static class OrderConstants
    {
        public const decimal ELECTRONICS_TAX_RATE = 0.1m;
        public const decimal BOOK_BULK_DISCOUNT_RATE = 0.05m;
        public const int BOOK_BULK_QUANTITY_THRESHOLD = 5;
        public const decimal FREE_SHIPPING_THRESHOLD = 100.0m;
        public const decimal STANDARD_SHIPPING_COST = 10.0m;
        
        public const string ELECTRONICS_CATEGORY = "ELECTRONICS";
        public const string BOOKS_CATEGORY = "BOOKS";
    }

    public class OrderCalculator
    {
        /// <summary>
        /// Calculates the total order amount including taxes, discounts, and shipping
        /// </summary>
        /// <param name="orderItems">List of items in the order</param>
        /// <returns>Total order amount</returns>
        public decimal CalculateOrderTotal(List<OrderItem> orderItems)
        {
            ValidateOrderItems(orderItems);

            decimal subtotal = CalculateSubtotal(orderItems);
            decimal shippingCost = CalculateShippingCost(subtotal);
            decimal totalAmount = subtotal + shippingCost;

            return Math.Round(totalAmount, 2);
        }

        private void ValidateOrderItems(List<OrderItem> orderItems)
        {
            if (orderItems == null || !orderItems.Any())
            {
                throw new ArgumentException("Order must contain at least one item");
            }

            foreach (var item in orderItems)
            {
                if (item.Quantity <= 0)
                {
                    throw new ArgumentException($"Invalid quantity for item {item.ProductName}");
                }

                if (item.Price < 0)
                {
                    throw new ArgumentException($"Invalid price for item {item.ProductName}");
                }

                if (string.IsNullOrWhiteSpace(item.ProductName))
                {
                    throw new ArgumentException("Product name cannot be empty");
                }
            }
        }

        private decimal CalculateSubtotal(List<OrderItem> orderItems)
        {
            decimal subtotal = 0;

            foreach (var item in orderItems)
            {
                decimal itemTotal = CalculateItemTotal(item);
                subtotal += itemTotal;
            }

            return subtotal;
        }

        private decimal CalculateItemTotal(OrderItem item)
        {
            decimal baseTotal = item.TotalPrice;

            // Apply category-specific pricing rules
            switch (item.Category?.ToUpper())
            {
                case OrderConstants.ELECTRONICS_CATEGORY:
                    return ApplyElectronicsTax(baseTotal);
                
                case OrderConstants.BOOKS_CATEGORY:
                    return ApplyBookDiscount(baseTotal, item.Quantity);
                
                default:
                    return baseTotal;
            }
        }

        private decimal ApplyElectronicsTax(decimal amount)
        {
            return amount * (1 + OrderConstants.ELECTRONICS_TAX_RATE);
        }

        private decimal ApplyBookDiscount(decimal amount, int quantity)
        {
            if (quantity >= OrderConstants.BOOK_BULK_QUANTITY_THRESHOLD)
            {
                return amount * (1 - OrderConstants.BOOK_BULK_DISCOUNT_RATE);
            }
            
            return amount;
        }

        private decimal CalculateShippingCost(decimal subtotal)
        {
            return subtotal >= OrderConstants.FREE_SHIPPING_THRESHOLD 
                ? 0 
                : OrderConstants.STANDARD_SHIPPING_COST;
        }
    }

    // Clean User Service with proper separation of concerns
    public interface IUserRepository
    {
        Task SaveUserAsync(User user);
        Task<User> GetUserByEmailAsync(string email);
    }

    public interface IEmailService
    {
        Task SendWelcomeEmailAsync(string email, string name);
    }

    public class UserService
    {
        private readonly IUserRepository userRepository;
        private readonly IEmailService emailService;

        public UserService(IUserRepository userRepository, IEmailService emailService)
        {
            this.userRepository = userRepository ?? throw new ArgumentNullException(nameof(userRepository));
            this.emailService = emailService ?? throw new ArgumentNullException(nameof(emailService));
        }

        /// <summary>
        /// Registers a new user in the system
        /// </summary>
        /// <param name="name">User's full name</param>
        /// <param name="email">User's email address</param>
        /// <param name="age">User's age</param>
        /// <returns>Newly created user</returns>
        public async Task<User> RegisterUserAsync(string name, string email, int age)
        {
            await ValidateUserDoesNotExistAsync(email);

            var newUser = CreateUser(name, email, age);
            await SaveUserAsync(newUser);
            await SendWelcomeEmailAsync(newUser);

            return newUser;
        }

        private async Task ValidateUserDoesNotExistAsync(string email)
        {
            var existingUser = await userRepository.GetUserByEmailAsync(email);
            if (existingUser != null)
            {
                throw new InvalidOperationException($"User with email {email} already exists");
            }
        }

        private User CreateUser(string name, string email, int age)
        {
            try
            {
                return new User(name, email, age);
            }
            catch (ArgumentException ex)
            {
                throw new ArgumentException($"Invalid user data: {ex.Message}", ex);
            }
        }

        private async Task SaveUserAsync(User user)
        {
            try
            {
                await userRepository.SaveUserAsync(user);
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException("Failed to save user to database", ex);
            }
        }

        private async Task SendWelcomeEmailAsync(User user)
        {
            try
            {
                await emailService.SendWelcomeEmailAsync(user.Email, user.Name);
            }
            catch (Exception ex)
            {
                // Log error but don't fail registration
                Console.WriteLine($"Warning: Failed to send welcome email to {user.Email}: {ex.Message}");
            }
        }
    }

    // Demo program showing clean code in action
    class Program
    {
        static void Main()
        {
            Console.WriteLine("=== Clean Code Demo ===\n");

            DemonstrateCleanUserClass();
            DemonstrateCleanOrderCalculation();

            Console.WriteLine("\n=== Clean Code Benefits ===");
            Console.WriteLine("✓ Code is easy to read and understand");
            Console.WriteLine("✓ Functions have single responsibilities");
            Console.WriteLine("✓ Names clearly express intent");
            Console.WriteLine("✓ Error handling is explicit and helpful");
            Console.WriteLine("✓ Constants replace magic numbers");
            Console.WriteLine("✓ Code is well-structured and organized");
        }

        private static void DemonstrateCleanUserClass()
        {
            Console.WriteLine("--- User Management Demo ---");

            try
            {
                // Create valid users
                var adultUser = new User("Alice Johnson", "alice@example.com", 25);
                var minorUser = new User("Bob Smith", "bob@example.com", 16);

                adultUser.IsVerified = true;

                adultUser.DisplayUserInfo();
                Console.WriteLine();
                minorUser.DisplayUserInfo();
                Console.WriteLine();
            }
            catch (ArgumentException ex)
            {
                Console.WriteLine($"❌ Error creating user: {ex.Message}");
            }

            // Demonstrate validation
            try
            {
                var invalidUser = new User("X", "invalid-email", -5);
            }
            catch (ArgumentException ex)
            {
                Console.WriteLine($"✅ Validation working: {ex.Message}");
            }
        }

        private static void DemonstrateCleanOrderCalculation()
        {
            Console.WriteLine("\n--- Order Calculation Demo ---");

            var orderItems = new List<OrderItem>
            {
                new OrderItem { ProductName = "Laptop", Price = 999.99m, Quantity = 1, Category = "ELECTRONICS" },
                new OrderItem { ProductName = "Programming Books", Price = 29.99m, Quantity = 6, Category = "BOOKS" },
                new OrderItem { ProductName = "Mouse Pad", Price = 9.99m, Quantity = 2, Category = "ACCESSORIES" }
            };

            var calculator = new OrderCalculator();
            decimal total = calculator.CalculateOrderTotal(orderItems);

            Console.WriteLine("Order Summary:");
            foreach (var item in orderItems)
            {
                Console.WriteLine($"  {item.ProductName}: ${item.Price} x {item.Quantity} = ${item.TotalPrice}");
            }
            Console.WriteLine($"\nTotal Amount: ${total}");
        }
    }
}